crates/cranelift/src/func_environ.rs

```diff
@@ -1115,10 +1127,146 @@ impl<'a, 'func, 'module_env> Call<'a, 'func, 'module_env> {
         callee: ir::Value,
         call_args: &[ir::Value],
     ) -> WasmResult<Option<ir::Inst>> {
+        // If we are performing call-indirect caching with this table, check the cache.
+        let caching = if self.env.tunables.cache_call_indirects {
+            let plan = &self.env.module.table_plans[table_index];
+            // We can do the indireaching optimization only if table
+            // elements will not change (no opcodes exist that could
+            // write the table, and table not exported), and if we can
+            // use the zero-index as a sentinenl for "no cache entry"
+            // (initial zeroed vmctx state).
+            !plan.written && !plan.non_null_zero
+        } else {
+            false
+        };
+
+        let (code_ptr, callee_vmctx) = if caching {
+            let index = self.env.call_indirect_index;
+            self.env.call_indirect_index += 1;
+            let vmctx = self.env.vmctx(self.builder.func);
+            let vmctx = self
+                .builder
+                .ins()
+                .global_value(self.env.pointer_type(), vmctx);
+            let offset = self
+                .env
+                .offsets
+                .vmctx_call_indirect_cache(u32::try_from(index).unwrap());
+            let call_indirect_cache_ptr = self.builder.ins().iadd_imm(vmctx, i64::from(offset));
+
+            let cached_index = self.builder.ins().load(
+                I32,
+                MemFlags::trusted(),
+                call_indirect_cache_ptr,
+                Offset32::from(self.env.offsets.ptr.vmcall_indirect_cache_index()),
+            );
+            let hit = self.builder.ins().icmp(IntCC::Equal, cached_index, callee);
+
+            let current_block = self.builder.current_block().unwrap();
+            let hit_block = self.builder.create_block();
+            let miss_block = self.builder.create_block();
+            self.builder.insert_block_after(hit_block, current_block);
+            self.builder.insert_block_after(miss_block, hit_block);
+            self.builder.set_cold_block(miss_block);
+            let continue_block = self.builder.create_block();
+            self.builder.insert_block_after(continue_block, miss_block);
```

Comment:
Could we factor this one out into smaller helpers and add comments describing the code and blocks we are emitting as well? I'm finding it really hard to read this giant block of clif-building code directly. See for example the overview comment and comments for each block we generate here:

https://github.com/bytecodealliance/wasmtime/blob/0f4c0d4a358187f99f23897d306c8b2a3c94287e/crates/cranelift/src/gc/enabled.rs#L341-L449

Reply:
Added a similar comment and factored out a little bit of the logic (and reduced the CFG by one block while I was at it) -- hopefully clearer now!